File: src-tauri/src/exa.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use tracing::{error, warn};

use crate::error::{self, AppError};

const EXA_SEARCH_URL: &str = "https://api.exa.ai/search";
const MAX_QUERY_LENGTH: usize = 2_000;
const MAX_NUM_RESULTS: u32 = 100;
const ERROR_BODY_MAX_LEN: usize = 200;
/// Maximum response body size (5 MiB). Prevents OOM from oversized API responses.
const MAX_RESPONSE_BYTES: usize = 5 * 1024 * 1024;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchRequest {
    pub query: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<SearchType>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<SearchCategory>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub num_results: Option<u32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub contents: Option<ContentOptions>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum SearchType {
    Neural,
    Fast,
    Auto,
    Deep,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum SearchCategory {
    #[serde(rename = "company")]
    Company,
    #[serde(rename = "research paper")]
    ResearchPaper,
    #[serde(rename = "news")]
    News,
    #[serde(rename = "pdf")]
    Pdf,
    #[serde(rename = "github")]
    Github,
    #[serde(rename = "tweet")]
    Tweet,
    #[serde(rename = "personal site")]
    PersonalSite,
    #[serde(rename = "financial report")]
    FinancialReport,
    #[serde(rename = "people")]
    People,
}

#[derive(Debug, Serialize)]
pub struct ContentOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<bool>,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchResponse {
    pub results: Vec<SearchResult>,
    pub request_id: Option<String>,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchResult {
    pub title: Option<String>,
    pub url: String,
    pub published_date: Option<String>,
    pub author: Option<String>,
    pub text: Option<String>,
    pub highlights: Option<Vec<String>>,
    pub score: Option<f64>,
    pub id: String,
}

pub struct ExaClient<'a> {
    http: &'a Client,
    api_key: &'a str,
}

impl<'a> ExaClient<'a> {
    pub fn new(http: &'a Client, api_key: &'a str) -> Self {
        Self { http, api_key }
    }

    pub async fn search(&self, request: &SearchRequest) -> Result<SearchResponse, AppError> {
        let response = self.send_request(request).await?;
        let status = response.status();

        if !status.is_success() {
            return Err(Self::classify_error_status(response).await);
        }

        Self::parse_response(response).await
    }

    async fn send_request(
        &self,
        request: &SearchRequest,
    ) -> Result<reqwest::Response, AppError> {
        self.http
            .post(EXA_SEARCH_URL)
            .header("x-api-key", self.api_key)
            .json(request)
            .send()
            .await
            .map_err(|e| {
                error::log_transport_error("Exa", &e);
                AppError::ExaRequest
            })
    }

    async fn classify_error_status(response: reqwest::Response) -> AppError {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        let safe_body = error::sanitize_error_body(&body, ERROR_BODY_MAX_LEN);
        warn!(status = %status, body = %safe_body, "Exa API error");

        match status.as_u16() {
            401 => AppError::ExaAuth,
            429 => AppError::ExaRateLimit,
            _ => AppError::ExaRequest,
        }
    }

    async fn parse_response(response: reqwest::Response) -> Result<SearchResponse, AppError> {
        if let Some(len) = response.content_length() {
            if len > MAX_RESPONSE_BYTES as u64 {
                error!(size = len, "Exa: response Content-Length exceeds size limit");
                return Err(AppError::ExaRequest);
            }
        }
        let bytes = response
            .bytes()
            .await
            .map_err(|_| {
                error!("Exa: failed to read response body");
                AppError::ExaRequest
            })?;
        if bytes.len() > MAX_RESPONSE_BYTES {
            error!(size = bytes.len(), "Exa: response body exceeds size limit");
            return Err(AppError::ExaRequest);
        }
        serde_json::from_slice::<SearchResponse>(&bytes).map_err(|_| {
            error!("Exa: failed to deserialize search response");
            AppError::ExaRequest
        })
    }
}

pub fn validate_search_request(request: &SearchRequest) -> Result<(), AppError> {
    if request.query.trim().is_empty() {
        return Err(AppError::Validation(
            "Search query must not be empty".into(),
        ));
    }
    if request.query.len() > MAX_QUERY_LENGTH {
        return Err(AppError::Validation(format!(
            "Search query exceeds maximum length of {} characters",
            MAX_QUERY_LENGTH
        )));
    }
    if let Some(n) = request.num_results {
        if n == 0 || n > MAX_NUM_RESULTS {
            return Err(AppError::Validation(format!(
                "numResults must be between 1 and {}",
                MAX_NUM_RESULTS
            )));
        }
    }
    Ok(())
}

File: apps/worker/src/exa.ts
import { HTTPException } from "hono/http-exception";
import type {
  SearchCategory,
  SearchRequest,
  SearchResponse,
  SearchResult,
  SearchType,
} from "./types";

const EXA_SEARCH_URL = "https://api.exa.ai/search";
const MAX_QUERY_LENGTH = 2000;
const MAX_NUM_RESULTS = 100;
const MAX_RESPONSE_BYTES = 5 * 1024 * 1024; // 5 MiB

const VALID_SEARCH_TYPES: ReadonlySet<string> = new Set<SearchType>([
  "neural",
  "fast",
  "auto",
  "deep",
]);

const VALID_CATEGORIES: ReadonlySet<string> = new Set<SearchCategory>([
  "company",
  "research paper",
  "news",
  "pdf",
  "github",
  "tweet",
  "personal site",
  "financial report",
  "people",
]);

function badRequest(message: string): never {
  throw new HTTPException(400, { message });
}

function validateQuery(value: unknown): string {
  if (typeof value !== "string") {
    badRequest("query must be a string");
  }
  const trimmed = value.trim();
  if (trimmed.length === 0) {
    badRequest("Search query must not be empty");
  }
  if (trimmed.length > MAX_QUERY_LENGTH) {
    badRequest(
      `Search query exceeds maximum length of ${MAX_QUERY_LENGTH} characters`
    );
  }
  return trimmed;
}

function validateNumResults(value: unknown): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== "number" || !Number.isInteger(value)) {
    badRequest("numResults must be an integer");
  }
  if (value < 1 || value > MAX_NUM_RESULTS) {
    badRequest(`numResults must be between 1 and ${MAX_NUM_RESULTS}`);
  }
  return value;
}

function validateEnum<T extends string>(
  value: unknown,
  fieldName: string,
  allowed: ReadonlySet<string>
): T | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== "string" || !allowed.has(value)) {
    badRequest(`${fieldName} must be one of: ${[...allowed].join(", ")}`);
  }
  return value as T;
}

function validateContents(value: unknown): { text?: boolean } | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    badRequest("contents must be an object");
  }
  const raw = value as Record<string, unknown>;
  if (raw.text !== undefined && typeof raw.text !== "boolean") {
    badRequest("contents.text must be a boolean");
  }
  return raw.text !== undefined ? { text: raw.text as boolean } : {};
}

export function validateSearchRequest(body: unknown): SearchRequest {
  if (body === null || typeof body !== "object" || Array.isArray(body)) {
    badRequest("Request body must be a JSON object");
  }

  const raw = body as Record<string, unknown>;

  const clean: SearchRequest = {
    query: validateQuery(raw.query),
  };

  const numResults = validateNumResults(raw.numResults);
  if (numResults !== undefined) {
    clean.numResults = numResults;
  }

  const type = validateEnum<SearchType>(raw.type, "type", VALID_SEARCH_TYPES);
  if (type !== undefined) {
    clean.type = type;
  }

  const category = validateEnum<SearchCategory>(
    raw.category,
    "category",
    VALID_CATEGORIES
  );
  if (category !== undefined) {
    clean.category = category;
  }

  const contents = validateContents(raw.contents);
  if (contents !== undefined) {
    clean.contents = contents;
  }

  return clean;
}

function badUpstream(): never {
  throw new HTTPException(502, { message: "Unexpected response from Exa" });
}

function validateSearchResult(item: unknown): SearchResult {
  if (item === null || typeof item !== "object" || Array.isArray(item)) {
    badUpstream();
  }

  const r = item as Record<string, unknown>;

  if (typeof r.id !== "string" || typeof r.url !== "string") {
    badUpstream();
  }

  return {
    id: r.id,
    url: r.url,
    title: typeof r.title === "string" ? r.title : null,
    publishedDate: typeof r.publishedDate === "string" ? r.publishedDate : null,
    author: typeof r.author === "string" ? r.author : null,
    text: typeof r.text === "string" ? r.text : null,
    highlights: Array.isArray(r.highlights)
      ? r.highlights.filter((h): h is string => typeof h === "string")
      : null,
    score: typeof r.score === "number" ? r.score : null,
  };
}

function validateSearchResponse(data: unknown): SearchResponse {
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    badUpstream();
  }

  const raw = data as Record<string, unknown>;

  if (!Array.isArray(raw.results)) {
    badUpstream();
  }

  return {
    results: raw.results.map(validateSearchResult),
    requestId: typeof raw.requestId === "string" ? raw.requestId : null,
  };
}

const EXA_FETCH_TIMEOUT_MS = 15_000;

async function fetchFromExa(
  apiKey: string,
  request: SearchRequest
): Promise<Response> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), EXA_FETCH_TIMEOUT_MS);

  try {
    return await fetch(EXA_SEARCH_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": apiKey,
      },
      body: JSON.stringify(request),
      signal: controller.signal,
    });
  } catch (err) {
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new HTTPException(504, { message: "Exa request timed out" });
    }
    throw new HTTPException(502, { message: "Exa search request failed" });
  } finally {
    clearTimeout(timeout);
  }
}

function handleExaErrorStatus(status: number): never {
  if (status === 401) {
    throw new HTTPException(401, { message: "Invalid Exa API key" });
  }
  if (status === 429) {
    throw new HTTPException(429, { message: "Exa rate limit exceeded" });
  }
  throw new HTTPException(502, { message: "Exa search request failed" });
}

function throwResponseTooLarge(): never {
  throw new HTTPException(502, {
    message: "Exa response exceeded size limit",
  });
}

async function readResponseBody(response: Response): Promise<unknown> {
  const contentLength = response.headers.get("Content-Length");
  if (
    contentLength &&
    Number.parseInt(contentLength, 10) > MAX_RESPONSE_BYTES
  ) {
    throwResponseTooLarge();
  }

  const text = await response.text();
  if (text.length > MAX_RESPONSE_BYTES) {
    throwResponseTooLarge();
  }

  try {
    return JSON.parse(text);
  } catch {
    badUpstream();
  }
}

export async function searchExa(
  apiKey: string,
  request: SearchRequest
): Promise<SearchResponse> {
  const response = await fetchFromExa(apiKey, request);

  if (!response.ok) {
    handleExaErrorStatus(response.status);
  }

  const parsed = await readResponseBody(response);
  return validateSearchResponse(parsed);
}
